src/librustc/ty/print/pretty.rs

```diff
@@ -988,16 +1014,33 @@ pub trait PrettyPrinter<'tcx>:
                 };
                 p!(print_value_path(instance.def_id(), instance.substs));
             }
-            // For zsts just print their type as their value gives no extra information
-            (Scalar::Raw { size: 0, .. }, _) => p!(print(ty)),
+            // For function type zsts just printing the type is enough
+            (Scalar::Raw { size: 0, .. }, ty::FnDef(..)) => p!(print(ty)),
+            // Empty tuples are frequently occurring, so don't print the fallback.
+            (Scalar::Raw { size: 0, .. }, ty::Tuple(ts)) if ts.is_empty() => p!(write("()")),
+            // Zero element arrays have a trivial representation.
+            (
+                Scalar::Raw { size: 0, .. },
+                ty::Array(
+                    _,
+                    ty::Const {
+                        val: ty::ConstKind::Value(ConstValue::Scalar(Scalar::Raw { data: 0, .. })),
+                        ..
+                    },
```

Comment:
Can't this be something like `if len.maybe_as_usize() == Some(0)`?

Reply:
The only convenience function we have is `len.try_eval_usize(self.tcx())`